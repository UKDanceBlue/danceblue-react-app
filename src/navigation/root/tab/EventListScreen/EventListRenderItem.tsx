import { useNavigation } from "@react-navigation/native";
import { FirestoreEvent } from "@ukdanceblue/db-app-common";
import { Interval } from "luxon";
import { Pressable } from "native-base";
import { MutableRefObject } from "react";
import { ListRenderItem } from "react-native";

import EventRow from "../../../../common/components/EventRow";
import { timestampToDateTime } from "../../../../common/util/dateTools";

import { dateFormat } from "./utils";

export const EventListRenderItem = ({
  item: thisEvent, index, dayIndexesRef
}: Parameters<ListRenderItem<FirestoreEvent>>[0] & { dayIndexesRef: MutableRefObject<Partial<Record<string, number>>> }) => {
  const { navigate } = useNavigation();

  if (thisEvent.interval != null) {
    const eventDate = timestampToDateTime(thisEvent.interval.start).toFormat(dateFormat);
    if (!((dayIndexesRef.current[eventDate] ?? NaN) > index)) {
      dayIndexesRef.current[eventDate] = index;
    }
  }
  return (
    <Pressable
      _pressed={{ opacity: 0.5 }}
      onPress={() => navigate("Event", { event: thisEvent })}
    >
      <EventRow
        title={thisEvent.name}
        blurb={thisEvent.shortDescription}
        imageSource={thisEvent.images?.[0]}
        interval={thisEvent.interval ? Interval.fromDateTimes(timestampToDateTime(thisEvent.interval.start), timestampToDateTime(thisEvent.interval.end)).toISO() : undefined}
      />
    </Pressable>
  );
};

export const makeEventListRenderItem = (dayIndexesRef: MutableRefObject<Partial<Record<string, number>>>): ListRenderItem<FirestoreEvent> => (props: Parameters<ListRenderItem<FirestoreEvent>>[0]) => EventListRenderItem({ ...props, dayIndexesRef });

