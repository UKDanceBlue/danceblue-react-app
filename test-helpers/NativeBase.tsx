import { render } from "@testing-library/react-native";
import { NativeBaseProvider, NativeBaseProviderProps } from "native-base";
import { ReactElement } from "react";

const inset = {
  frame: { x: 0, y: 0, width: 0, height: 0 },
  insets: { top: 0, left: 0, right: 0, bottom: 0 },
} as const;

/**
 * Editing this function will require an update to test snapshots.
 * @param element The element to test.
 * @param nativeBaseProps
 * @returns The output of *render()*, wrapped with the NativeBaseProvider.
 */
export const renderWithNativeBase = (element: ReactElement, nativeBaseProps: NativeBaseProviderProps = {}) => {
  return render(
    <NativeBaseProvider
      initialWindowMetrics={inset}
      colorModeManager={nativeBaseProps.colorModeManager}
      disableContrastText={nativeBaseProps.disableContrastText}
      theme={nativeBaseProps.theme}
      config={nativeBaseProps.config}
      isSSR={nativeBaseProps.isSSR}
    >
      {element}
    </NativeBaseProvider>
  );
};
