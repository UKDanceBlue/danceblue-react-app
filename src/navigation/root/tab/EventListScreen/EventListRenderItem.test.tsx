import { fireEvent } from "@testing-library/react-native";
import { FirestoreEvent } from "@ukdanceblue/db-app-common";

import { renderWithNativeBase } from "../../../../common/test-helpers/NativeBase";

import { EventListRenderItem } from "./EventListRenderItem";

describe("<EventListRenderItem />", () => {
  const testEvent = new FirestoreEvent(
    "test",
    "test",
    "test"
  );
  const tryToNavigate = jest.fn();

  const tree = renderWithNativeBase(<EventListRenderItem
    dayIndexesRef={{ current: {} }}
    index={0}
    item={testEvent}
    tryToNavigate={tryToNavigate}
  />);

  it("calls tryToNavigate when pressed", () => {
    expect(tree.queryByTestId("event-list-row-0-touchable")).not.toBeNull();
    fireEvent.press(tree.queryByTestId("event-list-row-0-touchable"));
    expect(tryToNavigate).toHaveBeenCalledWith(testEvent);
  });
});
