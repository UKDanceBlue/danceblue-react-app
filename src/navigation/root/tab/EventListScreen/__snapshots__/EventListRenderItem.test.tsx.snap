// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventListRenderItem /> renders correctly 1`] = `null`;
